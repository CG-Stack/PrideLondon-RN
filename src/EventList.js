// @flow
import React from "react";
import { View, StyleSheet, FlatList } from "react-native";
import EventCard from "./EventCard";
import type { Event } from "./integrations/cms";

type Props = {
  locale: string,
  events: Event[],
  refreshing: boolean,
  onRefresh: () => void
};

const EventList = ({ events, locale, refreshing, onRefresh }: Props) => (
  <FlatList
    data={events}
    keyExtractor={event => event.sys.id}
    renderItem={({ item: event }) => (
      <View style={styles.eventListItem}>
        <EventCard name={event.fields.name[locale]} />
      </View>
    )}
    refreshing={refreshing}
    onRefresh={onRefresh}
  />
);

const styles = StyleSheet.create({
  eventListItem: {
    paddingLeft: 15,
    paddingRight: 15,
    paddingBottom: 10
  }
});

export default EventList;
