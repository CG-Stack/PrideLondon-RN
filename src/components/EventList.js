// @flow
import React from "react";
import { StyleSheet, SectionList, TouchableOpacity, View } from "react-native";
import type { SectionBase } from "react-native/Libraries/Lists/SectionList";
import formatDate from "date-fns/format";

import EventCard from "./EventCard";
import Text from "./Text";
import type { Event, EventDays, Asset } from "../data/event";
import {
  bgColor,
  eventCardTextColor,
  sectionHeaderShadow,
  sectionHeaderBgColor,
  eventCardShadow
} from "../constants/colors";

type Props = {
  locale: string,
  events: EventDays,
  refreshing?: boolean,
  onRefresh?: () => void,
  onPress: (eventName: string) => void,
  getAssetById: string => Asset
};

const separator = style => () => <View style={style} />;

type ItemProps = { item: Event };
const renderItem = (styles, locale, onPress, getAssetById) => ({
  item: event
}: ItemProps) => (
  <TouchableOpacity
    style={styles.eventListItem}
    accessibilityTraits={["button"]}
    accessibilityComponentType="button"
    delayPressIn={50}
    onPress={() => onPress(event.sys.id)}
  >
    <EventCard
      name={event.fields.name[locale]}
      locationName={event.fields.locationName[locale]}
      eventPriceLow={event.fields.eventPriceLow[locale]}
      eventPriceHigh={event.fields.eventPriceHigh[locale]}
      startTime={event.fields.startTime[locale]}
      endTime={event.fields.endTime[locale]}
      imageUrl={`https:${
        getAssetById(event.fields.eventsListPicture[locale].sys.id).fields.file[
          locale
        ].url
      }`}
      isFree={event.fields.isFree[locale]}
    />
  </TouchableOpacity>
);

type Section = SectionBase<Event> & { title: string };

type SectionProps = { section: Section };
const renderSectionHeader = styles => ({ section }: SectionProps) => (
  <Text type="h2" style={styles.sectionHeader}>
    {section.title}
  </Text>
);

const eventSections = (events: EventDays, locale: string): Section[] =>
  events.map(it => ({
    data: it,
    title: formatDate(it[0].fields.startTime[locale], "dddd D MMMM")
  }));

const EventList = ({
  events,
  locale,
  refreshing,
  onRefresh,
  onPress,
  getAssetById
}: Props) => (
  <SectionList
    stickySectionHeadersEnabled
    sections={eventSections(events, locale)}
    renderSectionHeader={renderSectionHeader(styles)}
    renderSectionFooter={separator(styles.sectionFooter)}
    renderItem={renderItem(styles, locale, onPress, getAssetById)}
    keyExtractor={event => event.sys.id}
    contentContainerStyle={styles.container}
    ItemSeparatorComponent={separator(styles.itemSeparator)}
    SectionSeparatorComponent={separator(styles.sectionSeparator)}
    refreshing={refreshing}
    onRefresh={onRefresh}
  />
);

EventList.defaultProps = {
  refreshing: false,
  onRefresh: undefined
};

const styles = StyleSheet.create({
  itemSeparator: {
    height: 12
  },
  sectionSeparator: {
    height: 6
  },
  container: {
    paddingTop: 0,
    backgroundColor: bgColor
  },
  eventListItem: {
    marginHorizontal: 16,
    borderRadius: 5,
    // The below properties are required for ioS shadow
    shadowColor: eventCardShadow,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 1,
    shadowRadius: 3,
    // The below properties are required for android shadow
    borderWidth: 0,
    elevation: 3,
    backgroundColor: bgColor
  },
  sectionHeader: {
    height: 40,
    color: eventCardTextColor,
    paddingVertical: 8,
    paddingHorizontal: 15,
    backgroundColor: sectionHeaderBgColor,

    // The below properties are required for ioS shadow
    shadowColor: sectionHeaderShadow,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 1,
    shadowRadius: 3,
    // The below properties are required for android shadow
    borderWidth: 0,
    elevation: 3,
    marginBottom: 6
  },
  sectionFooter: {
    height: 6
  }
});

export default EventList;
