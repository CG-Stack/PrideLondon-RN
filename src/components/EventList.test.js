import React from "react";
import { shallow } from "enzyme";
import EventList from "./EventList";

const events = [
  [
    {
      sys: {
        id: "1",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some event"
        },
        startTime: {
          "en-GB": "2018-07-10T00:00"
        }
      }
    },
    {
      sys: {
        id: "2",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some other"
        },
        startTime: {
          "en-GB": "2018-07-10T00:00"
        }
      }
    }
  ],
  [
    {
      sys: {
        id: "2",
        type: "",
        contentType: { sys: { id: "" } },
        revision: 1
      },
      fields: {
        name: {
          "en-GB": "some other"
        },
        startTime: {
          "en-GB": "2018-07-11T00:00"
        }
      }
    }
  ]
];

describe("EventList", () => {
  const render = props =>
    shallow(
      <EventList
        events={events}
        locale="en-GB"
        refreshing={false}
        onRefresh={() => {}}
        onPress={() => {}}
        getAssetById={() => {}}
        {...props}
      />
    );

  it("renders correctly", () => {
    const output = render();
    expect(output).toMatchSnapshot();
  });

  describe("#shouldComponentUpdate", () => {
    const props = {
      locale: "en-GB",
      refreshing: false,
      events
    };

    it("stops update if locale, refresing and events stay the same", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: false,
        events: events.slice(0, 2) // force different instance
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(false);
    });

    it("allows update when locale changes", () => {
      const nextProps = {
        locale: "en-US",
        refreshing: false,
        events
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });

    it("allows update when refreshing", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: true,
        events
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });

    it("allows update when events changue", () => {
      const nextProps = {
        locale: "en-GB",
        refreshing: false,
        events: events.slice(0, 1)
      };

      const output = render(props);
      const shouldUpdate = output.instance().shouldComponentUpdate(nextProps);

      expect(shouldUpdate).toBe(true);
    });
  });
});
