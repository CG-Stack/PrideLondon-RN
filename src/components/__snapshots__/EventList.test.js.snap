// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<FlatList
  contentContainerStyle={
    Object {
      "paddingTop": 10,
    }
  }
  data={
    Array [
      Object {
        "fields": Object {
          "name": Object {
            "en-GB": "some event",
          },
        },
        "sys": Object {
          "contentType": Object {
            "sys": Object {
              "id": "",
            },
          },
          "id": "1",
          "revision": 1,
          "type": "",
        },
      },
      Object {
        "fields": Object {
          "name": Object {
            "en-GB": "some other",
          },
        },
        "sys": Object {
          "contentType": Object {
            "sys": Object {
              "id": "",
            },
          },
          "id": "2",
          "revision": 1,
          "type": "",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  onRefresh={[Function]}
  refreshing={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
